src/components/Tooltip/index.js

```diff
@@ -1,37 +1,239 @@
-    shouldRender: PropTypes.bool,
-};
+/**
+ * Choose the correct bounding box from the given list.
+ * This is a helper function for chooseBoundingBox below.
+ *
+ * @param {Element} bbs The bounding boxes of DOM element being hovered over.
+ * @param {number} clientX The X position from the MouseEvent.
+ * @param {number} clientY The Y position from the MouseEvent.
+ * @param {number} slop An allowed slop factor when searching for the bounding
+ * box. If the user is moving the mouse quickly we can end up getting a
+ * hover event with the position outside any of our bounding boxes. We retry
+ * with a small slop factor in that case, so if we have a bounding box close
+ * enough then we go with that.
+ * @return {DOMRect | null} The chosen bounding box. null if we failed to find
+ * a matching one, which can happen if the user is moving the mouse quickly
+ * and the onHoverOver event actually fires outside the element bounding box.
+ */
+function chooseBoundingBoxWithSlop(bbs, clientX, clientY, slop) {
+    for (let i = 0; i < bbs.length; i++) {
+        const bb = bbs[i];
+        if (bb.x - slop <= clientX && bb.x + bb.width + slop >= clientX && bb.y - slop <= clientY && bb.y + bb.height + slop >= clientY) {
+            return bb;
+        }
+    }
+    return null;
+}
+
+/**
+ * Choose the correct bounding box for the tooltip to be positioned against.
+ * This handles the case where the target is wrapped across two lines, and
+ * so we need to find the correct part (the one that the user is hovering
+ * over) and show the tooltip there.
+ *
+ * @param {Element} target The DOM element being hovered over.
+ * @param {number} clientX The X position from the MouseEvent.
+ * @param {number} clientY The Y position from the MouseEvent.
+ * @return {DOMRect} The chosen bounding box.
+ */
+function chooseBoundingBox(target, clientX, clientY) {
+    const bbs = target.getClientRects();
+    if (bbs.length === 1) {
+        return bbs[0];
+    }
+    let bb = chooseBoundingBoxWithSlop(bbs, clientX, clientY, 0);
+    if (bb) {
+        return bb;
+    }
+    // Retry with a slop factor, in case the user is moving the mouse quickly.
+    bb = chooseBoundingBoxWithSlop(bbs, clientX, clientY, 5);
+    if (bb) {
+        return bb;
+    }
+    // Fall back to the full bounding box if we failed to find a matching one.
+    // This could only happen if the user is moving the mouse very quickly
+    // and they got it outside our slop above.
+    return target.getBoundingClientRect();
+}
+
+/**
+ * A component used to wrap an element intended for displaying a tooltip. The term "tooltip's target" refers to the
+ * wrapped element, which, upon hover, triggers the tooltip to be shown.
+ * @param {propTypes} props
+ * @returns {ReactNodeLike}
+ */
+function Tooltip(props) {
+    const {children, numberOfLines, maxWidth, text, renderTooltipContent, renderTooltipContentKey} = props;
```

Comment:
Can't this be deconstructed above?

Reply:
It can but props was used with underscore. Ideally I think it's fine this way for now.